### ISSUE TITLE
Wind object from multiple NetCDF files

### ISSUE THREAD

**Issue by MBendoni** [NOAA-ORR-ERD/PyGnome#84]
Hi,

I would need to create a wind (and wave) object from multiple NetCDF files, to feed the weathering algorithm.
Concerning wind and current movers, I had no problem to create them by means a text file containing the path to the NetCDF files location:
`import gnome.scripting as gs`
`wind_file_list = gs.get_datafile(r'/myPath/wind_list')`
`wind_mover = GridWindMover(wind_file_list)`
`model.movers += wind_mover`
wind wind_list as follows:
NetCDF Files
[FILE] :/pathToFile/wind_2017011511.nc
[FILE] :/pathToFile/wind_2017011512.nc
..
..
And wind files containing time, lon, lat, air_u, air_v variables.

I would like to know how to perform something similar for wind object, in order to have a space-time variable wind (and wave) forcing for for the weathering algorithm.
I am able to do it only for a single file:
`wind_obj = GridWind.from_netCDF('/pathToFile/wind_2017011511.nc')`
`evaporation = Evaporation(water=water,wind=wind_obj)`
`model.weatherers += evaporation`

I tried also with the netCDF4.Dataset and get_dataset function but I think I am using a wrong syntax.

Any hint would be appreciated, thank you,
Michele

**Comment by ChrisBarker-NOAA** [comment]
The FileList method is the "old" way fo doing things. YOu should be able pass a list of filenames directly into the GridWind from_netCDF constructor:

```
wind_obj = GridWind.from_netCDF(['/pathToFile/wind_20170115.nc',
                                 '/pathToFile/wind_20170116.nc',
                                 '/pathToFile/wind_20170117.nc',
                                 ])
```

most of these will take either a single filename, a list of filenames, or a single netcdf Dataset.

**Comment by MBendoni** [comment]
Thank you, I tried and it works, even for the construction of wave object by means of wind oject.

Lastly, I would kindly ask for info about the way to refer to the pygnome package in papers, reports, etc...?
In addition, where can I find the references for the theretical framework at the basis of the weathering algorithm?

cheers,
Michele

**Comment by ChrisBarker-NOAA** [comment]
When passing in multiple netcdf files, it is using the netCDF4 MFDataset object:

https://unidata.github.io/netcdf4-python/

so very strange that it would not give the same errors in all uses.

If you can repeat the issue with stand alone code, we could take a look.

**Comment by ChrisBarker-NOAA** [comment]
> Lastly, I would kindly ask for info about the way to refer to the pygnome package in papers, reports, etc...

hmm -- now that I think about it, I'm not sure -- let me discuss here, and we'll let you know.

> In addition, where can I find the references for the theretical framework at the basis of the weathering algorithm?

Sorry about that -- we've been negligent in documenting all that properly. We have a draft technical document that is in pretty rough shape -- but does have most of it in there.

If you send a question to me at: Chris.Barker@noaa.gov -- I can send you a draft.

**Comment by MBendoni** [comment]
I tried to repeat the issue with the following code

`#!/usr/bin/env python`  
`from netCDF4 import MFDataset`
`a = MFDataset(['/OCEANASTORE/progetti/sharemed/oil_spill/tmp0.nc', '/OCEANASTORE/progetti/sharemed/oil_spill/tmp1.nc'])`
`print(ds.variables["time"][:])`

which gives:
[70176. 70177. 70178. 70179. 70180. 70181. 70182. 70183. 70184. 70185.
 70186. 70187. 70188. 70189. 70190. 70191. 70192. 70193. 70194. 70195.
 70196. 70197. 70198. 70199. 70176. 70177. 70178. 70179. 70180. 70181.
 70182. 70183. 70184. 70185. 70186. 70187. 70188. 70189. 70190. 70191.
 70192. 70193. 70194. 70195. 70196. 70197. 70198. 70199.]

It seems python  does not care that tmp0.nc has time units attribute:
"hours since 2009-1-1 00:00:00"
whereas tmp1.nc has:
"hours since 2009-1-2 00:00:00"
(there is a step between the 24-th and 25-th time output)

To clarify the aspect I report the email reply from Chris (thank you...):

_So that works, but you can see that there is a "jump" in the numbers from 70199 to  70176. I suspect the error will arise when we try to turn those into actual datetimes with the num2date function:
`netCDF4.num2date()` 
to use that, you need the units attribute, but it is not the same across the files, so a problem arises.
I just noticed that netCDF4 provides a MFTime which I think can help address this issue. 
I've posted an issue on the gridded project:
https://github.com/NOAA-ORR-ERD/gridded/issues/63
which we maintain -- it's used for most of the underlying handling of netcdf files, etc._

_CHB_


Michele

**Comment by ChrisBarker-NOAA** [comment]
I don't think this use case works yet (different time "epochs" in multiple files). but nor has anyone asked for it since then, so closing.
